### ISSUE TITLE
Action Required: Namespace migration from Microsoft.Web to Microsoft.App in March 2022

### ISSUE THREAD

**Issue by anthonychu** [microsoft/azure-container-apps#109]
---

**March 29, 2022 update:**
- Add Microsoft.App CLI installation instructions
- Azure portal now creates Microsoft.App resources by default

---

During the month of March 2022, all Azure Container Apps resources will be gradually migrated from the **Microsoft.Web** namespace to the **Microsoft.App** namespace.

**The availability of your Azure Container Apps resources will not be impacted.**

The migration affects how you issue management operations to Container Apps resources — the namespace you use _must_ match the resources you’re updating. If you're using the Azure CLI, you _must_ use a version of the Container Apps extension that is compatible with the resources you’re interacting with.

## <a name="migration"></a>Migration of existing Container Apps resources

Sometime during the migration period, your Container Apps resources that are in the Microsoft.Web namespace will be migrated to the Microsoft.App namespace. There may be a very short timeframe where Container Apps management requests are unavailable during the migration.

Following the migration, you’ll need to register your subscription with the Microsoft.App namespace before running management operations on your Container Apps resources.

```bash
az provider register -n Microsoft.App
```

### Azure portal

The Azure portal supports managing Microsoft.Web and Microsoft.App resources.

_(Update on March 29, 2022)_ - The Azure portal now creates Microsoft.App resources by default.

### ARM and Bicep

If you manage your Container Apps resources using ARM or Bicep templates or by calling ARM APIs, the namespace that you use must match the resources you’re interacting with. After the migration, you need to update your API calls and templates.

The changes for the Microsoft.App namespace are:

1. Update the API version to `2022-01-01-preview`
1. Update the `Microsoft.Web/containerApps` resource type to `Microsoft.App/containerApps`
1. Update the `Microsoft.Web/kubeEnvironments` resource type to `Microsoft.App/managedEnvironments`
1. Update the `kubeEnvironmentId` property to `managedEnvironmentId`

ARM template example:

```diff
{
  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "containerappName": {
      "defaultValue": "mycontainerapp",
      "type": "String"
    },
    "location": {
      "defaultValue": "canadacentral",
      "type": "String"
    },
    "environment_name": {
      "defaultValue": "myenvironment",
      "type": "String"
    }
  },
  "variables": {},
  "resources": [
    {
-      "apiVersion": "2021-03-01",
+      "apiVersion": "2022-01-01-preview",
-      "type": "Microsoft.Web/containerApps",
+      "type": "Microsoft.App/containerApps",
      "name": "[parameters('containerappName')]",
      "location": "[parameters('location')]",
      "properties": {
-        "kubeEnvironmentId": "[resourceId('Microsoft.Web/kubeEnvironments', parameters('environment_name'))]",
+        "managedEnvironmentId": "[resourceId('Microsoft.App/managedEnvironments', parameters('environment_name'))]",
        "configuration": {
          "secrets": [
            {
              "name": "mysecret",
              "value": "thisismysecret"
            }
          ],
          "ingress": {
            "external": true,
            "targetPort": 80,
            "allowInsecure": false,
            "traffic": [
              {
                "latestRevision": true,
                "weight": 100
              }
            ]
          }
        },
        "template": {
          "revisionSuffix": "myrevision",
          "containers": [
            {
              "name": "nginx",
              "image": "nginx",
              "env": [
                {
                  "name": "HTTP_PORT",
                  "value": "80"
                },
                {
                  "name": "SECRET_VAL",
                  "secretRef": "mysecret"
                }
              ],
              "resources": {
                "cpu": 0.5,
                "memory": "1Gi"
              }
            }
          ],
          "scale": {
            "minReplicas": 1,
            "maxReplicas": 3
          }
        }
      }
    }
  ]
}
```

For complete API specs for Microsoft.App, see this [GitHub repo](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/app/resource-manager).

### Azure CLI

The Azure Container Apps commands in the Azure CLI are currently enabled by a CLI extension. You must use an extension version that is compatible with the namespace of the Container Apps resources you want to interact with.

_(Update on March 29, 2022)_ - To manage resources in the Microsoft.App namespace, install the latest Azure Container Apps CLI extension. Note that the installation instructions have changed. You will not be able to use older version of the CLI extension to manage your resources after the migration.

```
# uninstall previous versions
az extension remove --name containerapp

# install the latest
az extension add --name containerapp
```

See https://github.com/microsoft/azure-container-apps/issues/152 for details on this CLI extension release.

### Pulumi

If you're using Pulumi, see their migration guide for more details: https://github.com/pulumi/pulumi/discussions/9099

## Creating new Container Apps resources

Until your Azure subscription has migrated to use the Microsoft.App namespace, you may continue to create Container App Environments in the Microsoft.Web namespace.

When creating new Container Apps, you must use the same namespace as the Container App Environment it’s being created in.

In early March 2022, new Azure Container App Environments created from the Azure portal will use the Microsoft.App namespace. You can manage both Microsoft.Web and Microsoft.App Container Apps resources in the Azure portal. However, when issuing management operations, you must ensure you’re using the correct namespace or a compatible Azure CLI extension version.

See the [Migration of existing Container Apps resources](#migration) section above for more details on what to do when your Container App Environments and Container Apps are migrated.

## Frequently asked questions

**Can you tell me when my resources will be updated to Microsoft.App?**

The migration is performed in batches and we’re unable to provide the exact migration timeframe for your resources.

**How will I know when my resources have been updated?**

After your resources have been migrated to Microsoft.App, operations using Microsoft.Web will fail with an error such as _“Creation or update of environments and container apps in Microsoft.Web namespace are no longer supported. Please use Microsoft.App namespace instead.”_

**Where can I ask questions related to this migration and get help?**

Respond to this issue or create a new issue in this repo. You can also open a case with Azure Support.

**Comment by mikhailshilkov** [comment]
When will the new specs land in https://github.com/Azure/azure-rest-api-specs? I hope it's well before the actual migration of resources, otherwise it's going to break our users of the Pulumi Azure Native provider.

**Comment by ThorstenHans** [comment]
@anthonychu will this also affect reading existing instances, so will the type of existing ACA instances be modified?

If that's the case, I've created a small GitHub Action that customers can use to get a notification once the migration has updated at least one existing instance of their Azure Container Apps

https://github.com/ThorstenHans/check-aca-arm-namespace-migration

If existing ACA instances were not updated in place, I've to update the Action once again 😄 However, I thought this could be useful. The repo also shows how to have that check being executed every 12 hrs using a cron schedule

**Comment by ruslany** [comment]
> When will the new specs land in https://github.com/Azure/azure-rest-api-specs? I hope it's well before the actual migration of resources, otherwise it's going to break our users of the Pulumi Azure Native provider.

@mikhailshilkov - The new version branch is ready to be merged to the main branch of azure-rest-api-spec. It should be merged pretty soon. Meanwhile you can find the AP specs here: 

https://github.com/Azure/azure-rest-api-specs/tree/Microsoft.App-2022-01-01-preview/specification/app/resource-manager

**Comment by ruslany** [comment]
> 

@ThorstenHans - thanks for preparing the github action for this! 

The way the migration works is that all subscription's environments and containerapps hosted there are migrated at the same time in a given region. After the migration the resources will not be visible in the Microsoft.Web namespace but instead will be visible in the Microsoft.App namespace. So I guess the action could be modified to use the new go client sdk (when it is available) together with the old client sdk to list environments in both namespaces.

**Comment by ThorstenHans** [comment]
@ruslany correct. That was the intention once the new SDK is there. 

With the current iteration, I can get an immediate notification when the first ACA is migrated in my sub.

Having both SDKs would allow to create a nice listing of all the ACA resources (both migrated and current ones under Microsoft.Web)

**Comment by ruslany** [comment]
@danielrbradley ,

Yes, there will be no transition period when resources are available in both namespaces. However, the portal will work seamlessly with both namespaces and you can continue using it before and after the migration without any noticeable difference. The reason the resources are not available for updates in both namespaces at the same time is because that would violate the ARM requirement that the resource state is updated only via the ARM API calls. For example if an app is created in Microsoft.Web namespace and then updated in the Microsoft.App namespace – the app’s state in the Microsoft.Web will change without any explicit actions performed on that app in that namespace. That will make the ARM cache, Azure Resource Graph and Activity log in Microsoft.Web namespace out of sync with the actual object state.

All resources within a subscription in a region will be migrated at once. But resources across regions may be migrated at different times.

The specifications for the new Microsoft.App namespace will be merged to the main branch of azure-rest-api-specs very soon. Meanwhile you can see them in a version branch here: [azure-rest-api-specs/specification/app/resource-manager at Microsoft.App-2022-01-01-preview · Azure/azure-rest-api-specs (github.com)](https://github.com/Azure/azure-rest-api-specs/tree/Microsoft.App-2022-01-01-preview/specification/app/resource-manager). The API version is still a preview version because we keep iterating on the APIs as new functionality is added. The existing API version from the Microsoft.Web namespace will not be transitioned to Microsoft.App.

**Comment by danielrbradley** [comment]
Thanks for the additional detail @ruslany 

Are you also notifying any active users of the service directly with details of how to migrate?

**Comment by ruslany** [comment]
@danielrbradley - Yes, we are planning to send notifications to all active users. The notification will reference this github issue which we will be updating with migration details.

**Comment by isaacabraham** [comment]
@ruslany What do you recommend authors of tools that use ARM do for our users e.g. I'm one of the maintains of [Farmer](https://compositionalit.github.io/farmer/) and we have had since last year a [working implementation](https://compositionalit.github.io/farmer/api-overview/resources/container-apps/) of container apps.

Obviously if there's a "hard cutover" then some of our users (I imagine this is a relatively small number, but still) will be left high and dry in some way:

* If we wait to release a fix in Farmer until _after_ the cut over, users will be stuck until then.
* If we release before a fix in Farmer until _before_ the cut over, users who prematurely upgrade will be stuck until the change over occurs.

Because Farmer removes the need to know how to generate the ARM (or which ARM resources paths are generated), people's Farmer code will be completely unchanged, but the ARM that's generated will suddenly start changing - an situation we're really careful to avoid.

The only other idea I have is a period of time whereby we have make users set flag to explicitly upgrade to the new schema, and then after a period of time, remove the flag and make this the default.

@forki I believe you might be affected by this. Also looping in @ninjarobot for any other ideas.

**Comment by ruslany** [comment]
@isaacabraham - the option where user sets a flag to use old or new schema sounds reasonable and should work. I am also wondering if instead of introducing a new flag, the decision to use old or new schema can be driven by what type of an environment is chosen by user. There will be no `kubeEnvironments` resource type in the `Microsoft.App` namespace. Instead there is `Microsoft.App/managedEnvironments` resource type. `The managedEnvironment` resource type does not exist in the `Microsoft.Web`. Would it make sense to introduce a ManagedEnvironment builder? When user starts with that then the new ARM schema is used for creating environments and the apps in it. If user starts with an existing Container Environment builder then the old Microsoft.Web schema is used. Then after the migration is done the existing Container Environment builder is deprecated.

**Comment by willvelida** [comment]
Hey @anthonychu, I'm in the process of migrating my Container Apps Environment and Container App to the new namespace in my Bicep templates and deploying it via GitHub Actions. On the Bicep validation step, I get the following error:

```
[validate](https://github.com/willvelida/bookstore-containerapps/runs/5383975234?check_suite_focus=true)
ERROR: {'code': 'InvalidTemplateDeployment', 'message': "The template deployment '18' is not valid according to the validation procedure. The tracking id is 'f99b8305-c219-4795-8c9b-eac864116bba'. See inner errors for details."}

Inner Errors: 
{'code': '90010', 'message': 'LogAnalyticsConfiguration.CustomerId is invalid.  CustomerId must be a GUID'}
```

Looking at the [API Definition](https://github.com/Azure/azure-rest-api-specs/blob/4042a47cff72c950d35dd72149895b044ed79713/specification/app/resource-manager/Microsoft.App/preview/2022-01-01-preview/ManagedEnvironments.json#L770), it looks like the type required for Log Analytics Customer Id is still a string, so has the API Definition been updated as part of this migration or is there something else going on? 

I've raised the following issue for more details: #126

**Comment by ruslany** [comment]
Hi @willvelida - this is a bug in the new PR implementation. We are deploying the fix for it in a next couple of days.

**Comment by danielrbradley** [comment]
For anyone using Pulumi to manage their Container Apps can follow this migration guide for managing the transition of resources between namespaces: https://github.com/pulumi/pulumi/discussions/9099

**Comment by willvelida** [comment]
@ruslany I've just updated my Bicep template, run my deployment and it works! 😊

I'm guessing this means that the fix has been deployed? If so, I can go ahead and close #126

**Comment by ruslany** [comment]
@willvelida - thanks for the verification! I was about to post here that we've deployed the fix :). You can close the #126

**Comment by takekazuomi** [comment]
When I deployed in the new namespace(Microsoft.App/managedEnvironments), I got the following error.

```
 "ContainerAppsConfiguration.DaprAIInstrumentationKey is invalid.  DaprAIInstrumentationKey length cannot be greater than 64 characters."
```

Curiously, neither LogAnalytics nor ApplicationInsights have been created after the error. These two should have been created before 'Microsoft.App/managedEnvironments' in the dependencies.

bicep code is here
https://github.com/takekazuomi/aca-bicep01/blob/main/deploy-env/environment.bicep

**Comment by davidxw** [comment]
I've deployed into the new namespace (using Bicep), and have noticed a couple of issues:

- Ingress property on bicep template seems to be now required, however it is not validated by schema, and deployment of the container app fails (template missing the ingress property deployed without issue when using Microsoft.Web)
- Logging/metrics broken in portal - error returned is "microsoft.app/containerapps is not a supported platform metric namespace, supported ones are Microsoft.AnalysisServices ...."

**Comment by timoknapp** [comment]
> When I deployed in the new namespace(Microsoft.App/managedEnvironments), I got the following error.
> 
> ```
>  "ContainerAppsConfiguration.DaprAIInstrumentationKey is invalid.  DaprAIInstrumentationKey length cannot be greater than 64 characters."
> ```
> 
> Curiously, neither LogAnalytics nor ApplicationInsights have been created after the error. These two should have been created before 'Microsoft.App/managedEnvironments' in the dependencies.
> 
> bicep code is here https://github.com/takekazuomi/aca-bicep01/blob/main/deploy-env/environment.bicep

I am having the exact same issue since I migrated the API from `Microsoft.Web` to `Microsoft.App`. Did you manage to resolve the issue @takekazuomi

**Comment by ruslany** [comment]
@takekazuomi , @timoknapp, @davidxw - thanks for reporting these bugs. We are looking into them. The metrics/logging being broken in portal is a known issue as the onboarding of Microsoft.App to metrics is still in progress.

**Comment by ruslany** [comment]
@takekazuomi , @timoknapp - the fix for the DaprAIInstrumentationKey error is being deployed and should be in all production regions shortly.

**Comment by ruslany** [comment]
@davidxw - what error are you getting when you try to deploy an app without ingress? I tried to repro this with the template that does not have ingress section but it went through without errors.

**Comment by KaiWalter** [comment]
I aligned my custom vnet template to <https://github.com/Azure/azure-rest-api-specs/blob/Microsoft.App-2022-01-01-preview/specification/app/resource-manager/Microsoft.App/preview/2022-01-01-preview/ManagedEnvironments.json>

```
resource environment 'Microsoft.App/managedEnvironments@2022-01-01-preview' = {
  name: environmentName
  location: location
  properties: {
    appLogsConfiguration: {
      destination: 'log-analytics'
      logAnalyticsConfiguration: {
        customerId: logAnalyticsCustomerId
        sharedKey: logAnalyticsSharedKey
      }
    }
    daprAIInstrumentationKey: appInsightsInstrumentationKey
    vnetConfiguration: {
      infrastructureSubnetId: '${vnetId}/subnets/cp'
      runtimeSubnetId: '${vnetId}/subnets/apps'
      internal: true
    }
  }
}
```

but after deployment the resource did not seem to pick up this configuration and has no vnet configuration:

```
> az resource show -g ca-kw -n ca-kw --resource-type Microsoft.App/managedEnvironments
{
  "extendedLocation": null,
  "id": "/subscriptions/**********************************/resourceGroups/ca-kw/providers/Microsoft.App/managedenvironments/ca-kw",
  "identity": null,
  "kind": null,
  "location": "westeurope",
  "managedBy": null,
  "name": "ca-kw",
  "plan": null,
  "properties": {
    "appLogsConfiguration": {
      "destination": "log-analytics",
      "logAnalyticsConfiguration": {
        "customerId": "de074225-6a64-4687-b94d-f112f071f06b"
      }
    },
    "defaultDomain": "purplepond-275aae86.westeurope.azurecontainerapps.io",
    "provisioningState": "Succeeded",
    "staticIp": "40.114.178.40"
  },
  "resourceGroup": "ca-kw",
  "sku": null,
  "systemData": {
    "createdAt": "2022-03-05T05:38:20.1890136",
    "createdBy": "A1C_Y1WLE@zeiss.com",
    "createdByType": "User",
    "lastModifiedAt": "2022-03-05T07:22:39.0511734",
    "lastModifiedBy": "A1C_Y1WLE@zeiss.com",
    "lastModifiedByType": "User"
  },
  "tags": null,
  "type": "Microsoft.App/managedenvironments"
}
```

is there an ETA when this is restored?
Thanks

**Comment by davidxw** [comment]
> @davidxw - what error are you getting when you try to deploy an app without ingress? I tried to repro this with the template that does not have ingress section but it went through without errors.

When I was attempting this late last week I was getting an internal server error, repeated on re-try. I had upgraded the [reddog/containerapps](https://github.com/Azure/reddog-containerapps) repo and the virtual-customer app failed deployment until I added an ingress config. I'll test again and post a repro.

**Comment by takekazuomi** [comment]
@ruslany DaprAIInstrumentationKey works fine now.
@KaiWalter Also vnetConfiguration work fine too.

```json
$ az resource show -g omi02-rg -n acaenv-omi02 --resource-type Microsoft.App/managedEnvironments
{
  "extendedLocation": null,
  "id": "/subscriptions/**********/resourceGroups/omi02-rg/providers/Microsoft.App/managedenvironments/acaenv-omi02",
  "identity": null,
  "kind": null,
  "location": "canadacentral",
  "managedBy": null,
  "name": "acaenv-omi02",
  "plan": null,
  "properties": {
    "appLogsConfiguration": {
      "destination": "log-analytics",
      "logAnalyticsConfiguration": {
        "customerId": "808d8386-78cf-4144-883b-cc3699c16ccd"
      }
    },
    "defaultDomain": "whitesand-980669c1.canadacentral.azurecontainerapps.io",
    "provisioningState": "Succeeded",
    "staticIp": "***.***.***.***",
    "vnetConfiguration": {
      "dockerBridgeCidr": "10.2.0.1/16",
      "infrastructureSubnetId": "/subscriptions/**********/resourceGroups/omi02-rg/providers/Microsoft.Network/virtualNetworks/vnet/subnets/control-plane",
      "internal": false,
      "platformReservedCidr": "10.1.0.0/16",
      "platformReservedDnsIP": "10.1.0.2",
      "runtimeSubnetId": "/subscriptions/**********/resourceGroups/omi02-rg/providers/Microsoft.Network/virtualNetworks/vnet/subnets/applications"
    }
  },
  "resourceGroup": "omi02-rg",
  "sku": null,
  "systemData": {
    "createdAt": "2022-03-05T08:43:20.2837628",
    "createdBy": "tomi@baleen.studio",
    "createdByType": "User",
    "lastModifiedAt": "2022-03-05T08:43:20.2837628",
    "lastModifiedBy": "tomi@baleen.studio",
    "lastModifiedByType": "User"
  },
  "tags": null,
  "type": "Microsoft.App/managedenvironments"
}
```

**Comment by davidxw** [comment]
> > @davidxw - what error are you getting when you try to deploy an app without ingress? I tried to repro this with the template that does not have ingress section but it went through without errors.
> 
> When I was attempting this late last week I was getting an internal server error, repeated on re-try. I had upgraded the [reddog/containerapps](https://github.com/Azure/reddog-containerapps) repo and the virtual-customer app failed deployment until I added an ingress config. I'll test again and post a repro.

Here is the full template that fails for me - issue is with the container app, not the container app environment:

`param location string = resourceGroup().location
param uniqueSeed string = '${subscription().subscriptionId}-${resourceGroup().name}'
param uniqueSuffix string = 'reddog-${uniqueString(uniqueSeed)}'
param containerAppsEnvName string = 'cae-${uniqueSuffix}'
param logAnalyticsWorkspaceName string = 'log-${uniqueSuffix}'
param appInsightsName string = 'appi-${uniqueSuffix}'

resource logAnalyticsWorkspace 'Microsoft.OperationalInsights/workspaces@2020-03-01-preview' = {
  name: logAnalyticsWorkspaceName
  location: location
  properties: any({
    retentionInDays: 30
    features: {
      searchVersion: 1
    }
    sku: {
      name: 'PerGB2018'
    }
  })
}

resource appInsights 'Microsoft.Insights/components@2020-02-02-preview' = {
  name: appInsightsName
  location: location
  kind: 'web'
  properties: { 
    Application_Type: 'web'
  }
}

resource containerAppsEnv 'Microsoft.App/managedEnvironments@2022-01-01-preview' = {
  name: containerAppsEnvName
  location: location
  properties: {
    type: 'managed'
    internalLoadBalancerEnabled: false
    appLogsConfiguration: {
      destination: 'log-analytics'
      logAnalyticsConfiguration: {
        customerId: logAnalyticsWorkspace.properties.customerId
        sharedKey: logAnalyticsWorkspace.listKeys().primarySharedKey
      }
    }
    containerAppsConfiguration: {
      daprAIInstrumentationKey: appInsights.properties.InstrumentationKey
    }
  }
}

resource virtualCustomers 'Microsoft.App/containerApps@2022-01-01-preview' = {
  name: 'virtual-customers'
  location: location
  properties: {
    managedEnvironmentId: containerAppsEnv.id
    template: {
      containers: [
        {
          name: 'virtual-customers'
          image: 'ghcr.io/azure/reddog-retail-demo/reddog-retail-virtual-customers:latest'
        }
      ]
      scale: {
        minReplicas: 1
      }
    }
  }
}
`

**Comment by davidxw** [comment]
I've also noticed that dapr doesn't seem to be working with this deployment, e.g. the container app deployed above has code to invoke another service via dapr:

[2022-03-06 12:11:50.681 ERR] [RedDog.VirtualCustomers.VirtualCustomers] Error retrieving products. Retrying in 5 seconds. Message: **Connection refused (127.0.0.1:3500)**

Is there anything else I can do to troubleshoot this, or is this also a known issue?

**Comment by KaiWalter** [comment]
ありがとうございました @takekazuomi - also my vnet configuration is working again

**Comment by timoknapp** [comment]
> I've also noticed that dapr doesn't seem to be working with this deployment, e.g. the container app deployed above has code to invoke another service via dapr:
> 
> [2022-03-06 12:11:50.681 ERR] [RedDog.VirtualCustomers.VirtualCustomers] Error retrieving products. Retrying in 5 seconds. Message: **Connection refused (127.0.0.1:3500)**
> 
> Is there anything else I can do to troubleshoot this, or is this also a known issue?

I have the same issue with Dapr since I upgraded to the new `Microsoft.App`API.

**Comment by KaiWalter** [comment]
same here with Dapr - 2 simple Apps deployed with Dapr
<https://github.com/KaiWalter/container-apps-experimental/blob/namespace-migration/app1/Controllers/HealthController.cs>
<https://github.com/KaiWalter/container-apps-experimental/blob/namespace-migration/ca-bicep/build-apps.sh>

apps can be called individually but they do not succeed with service invocation
<https://github.com/KaiWalter/container-apps-experimental/blob/namespace-migration/ca-bicep/test-apps.sh>

```
        [Route("health")]
        [HttpGet]
        public IActionResult Get()
        {
            var status = new
            {
                status = "OK",
                assembly = System.Reflection.Assembly.GetExecutingAssembly().FullName,
                instrumentationKey = Environment.GetEnvironmentVariable("INSTRUMENTATIONKEY"),
            };

            return Ok(status);
        }

        [Route("health-remote")]
        [HttpGet]
        public async Task<IActionResult> App2Get()
        {
            var status = await _daprClient.InvokeMethodAsync<object>(HttpMethod.Get, "app2", "health");

            return Ok(status);
        }
```

**Comment by ruslany** [comment]
@davidxw - thanks for posting the repro for internal server error when deploying an app without ingress. Found what was causing the error. The fix will be deployed in next couple of days.

**Comment by ruslany** [comment]
> ありがとうございました @takekazuomi - also my vnet configuration is working again

@KaiWalter - did you make any changes to get this working or you just retried and it worked?

**Comment by vinisoto** [comment]
> @takekazuomi , @timoknapp, @davidxw - thanks for reporting these bugs. We are looking into them. The metrics/logging being broken in portal is a known issue as the onboarding of Microsoft.App to metrics is still in progress.

Hi, metrics are now enabled for microsoft.app/containerapps

**Comment by takekazuomi** [comment]
Congratulations metrics 🎉
![image](https://user-images.githubusercontent.com/73158/157104219-5943fc28-8e36-4683-a66c-11cbc51f1cf1.png)

**Comment by ruslany** [comment]
@KaiWalter - can you try modifying the dapr section as below?

```
dapr: {
        enabled: true
        appId: name
        appPort: containerPort
        appProtocol: 'http'
      }
```

**Comment by timoknapp** [comment]
> @KaiWalter - can you try modifying the dapr section as below?
> 
> 
> 
> ```
> 
> dapr: {
> 
>         enabled: true
> 
>         appId: name
> 
>         appPort: containerPort
> 
>         appProtocol: 'http'
> 
>       }
> 
> ```

@ruslany Can you maybe provide a more comprehensive sample? Cause your snippet looks pretty much like [this one](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/app/resource-manager/Microsoft.App/preview/2022-01-01-preview/examples/ContainerApps_CreateOrUpdate.json#L28:L32) but Dapr still does not work. Furthermore, it seems that after provisioning container apps the related Dapr configuration is not part of overall configuration anymore.

**Comment by clarenceb** [comment]
Has anyone managed to get the daprComponents resource working?  I have tried a few different ways but I keep getting Bad Request errors.  My snippet look like this:

```
resource environment 'Microsoft.App/managedEnvironments@2022-01-01-preview' existing = {
  name: environmentName
}

resource statestore 'Microsoft.App/managedEnvironments/daprComponents@2022-01-01-preview' = {
  name: 'statestore'
  parent: environment
  
  properties: {
    type: 'state.azure.blobstorage'
    version: 'v1'

    secrets: [
      {
        name: 'storage-key'
        value: blobstore.listKeys().keys[0].value
      }
    ]

    metadata: [
        {
            name: 'accountName'
            value: blobstore.name
        }
        {
            name: 'accountKey'
            secretRef: 'storage-key'
        }
        {
            name: 'containerName'
            value: storageAccountName
        }
    ]

    scopes: [
      'nodeapp'
    ]
  }
}
```

**Comment by ruslany** [comment]
Hello @clarenceb - the daprcomponents API implementation is not deployed to production yet. We expect it to be available in a week.

**Comment by ruslany** [comment]
@timoknapp - please hold off on testing the dapr functionality. We are making some changes to the API specification and implementation. We'll update the announcement with more details shortly.

**Comment by clarenceb** [comment]
@ruslany I see...well that explains why it's not working.  Please update here when it's released so we can test it out.

**Comment by KaiWalter** [comment]
> @KaiWalter - can you try modifying the dapr section as below?
> 
> ```
> dapr: {
>         enabled: true
>         appId: name
>         appPort: containerPort
>         appProtocol: 'http'
>       }
> ```

Thanks @ruslany - that was the missing ingredients - sorry I overlooked this

```
app1
{"status":"OK","assembly":"app1, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","instrumentationKey":null} <<-- check app1 own health
{"status":"OK","assembly":"app2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","instrumentationKey":null} <<-- check app1 remote health
app2
{"status":"OK","assembly":"app2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","instrumentationKey":null} <<-- check app2 own health
{"status":"OK","assembly":"app1, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","instrumentationKey":null} <<-- check app2 remote health
```

**Comment by KaiWalter** [comment]
> > ありがとうございました @takekazuomi - also my vnet configuration is working again
> 
> @KaiWalter - did you make any changes to get this working or you just retried and it worked?

@ruslany I just redeployed my already fixed / migrated configuration
<https://github.com/KaiWalter/container-apps-experimental/blob/main/ca-bicep/environment.bicep>

**Comment by kendallroden** [comment]
Hi all! Thanks for this feedback. Please keep in mind that we are in the middle of making several dapr-related geared at enhancing your overall dapr experience. There are changes in flight, and there should be more alignment next week. You can continue to leverage the .Web namespace to run your dapr demos and scenarios. Happy to provide additional clarity if needed!

**Comment by ruslany** [comment]
Hello all, this announcement describes the upcoming Dapr functionality changes: 

[IMPORTANT: Dapr configuration changes for Container Apps](https://github.com/microsoft/azure-container-apps/issues/132)

**Comment by simongullberg** [comment]
Will the dapr version running in ACA be updated to a newer version when migrating to Microsoft.App namespace?

**Comment by ruslany** [comment]
@simongullberg - The dapr version will not be updated as part of migration to Microsoft.App. We're tracking a dapr version upgrade to the latest version sometime after the migration.

**Comment by lniguidula** [comment]
Hi everyone, I tried to deploy an ACA via ARM using Microsoft.Web but we run into some connection issues when accessing the app. We're not sure if this is related to the migration since we haven't started the changes on our end.

**Comment by adeturner** [comment]
Hi, any update on the new AZ CLI version?

```
az containerapp env create -n $ENV -g $RG --l $LOC --logs-workspace-id $LOG_AW_ID --logs-workspace-key $LOG_AW_KEY 

Creation or update of environments and container apps in Microsoft.Web namespace are no longer supported. 
Please use Microsoft.App namespace instead. More information available via aka.ms/microsoftappmigration.
```

https://github.com/microsoft/azure-container-apps/issues/109

The Azure Container Apps commands in the Azure CLI are currently enabled by a CLI extension. 
You must use an extension version that is compatible with the namespace of the Container Apps resources you want to interact with.
**We'll update this announcement when the CLI extension version for working with Microsoft.App is available.**

**Comment by adeturner** [comment]
Just to add, looks like I cant deploy via the portal or az deployment template either...

With "type": "Microsoft.Web/kubeEnvironments":

```
az deployment group create --resource-group ${RG} --name rollout01 --template-file arm/containerapp-env.json --parameters @containerapp-env-params.json
{"status":"Failed","error":{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"Conflict","message":"{\r\n  \"Code\": \"90022\",\r\n  \"Message\": \"Creation or update of environments and container apps in Microsoft.Web namespace are no longer supported. Please use Microsoft.App namespace instead. More information available via aka.ms/microsoftappmigration.\",\r\n  \"Target\": null,\r\n  \"Details\": [\r\n    {\r\n      \"Message\": \"Creation or update of environments and container apps in Microsoft.Web namespace are no longer supported. Please use Microsoft.App namespace instead. More information available via aka.ms/microsoftappmigration.\"\r\n    },\r\n    {\r\n      \"Code\": \"90022\"\r\n    },\r\n    {\r\n      \"ErrorEntity\": {\r\n        \"Code\": \"90022\",\r\n        \"Message\": \"Creation or update of environments and container apps in Microsoft.Web namespace are no longer supported. Please use Microsoft.App namespace instead. More information available via aka.ms/microsoftappmigration.\"\r\n      }\r\n    }\r\n  ],\r\n  \"Innererror\": null\r\n}"}]}}
```

With "type": "Microsoft.App/kubeEnvironments"

```
$ az deployment group create --resource-group $RG --name rollout01 --template-file arm/containerapp-env.json --parameters @containerapp-env-params.json

{"status":"Failed","error":{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"NotFound","message":"{\r\n  \"error\": {\r\n    \"code\": \"InvalidResourceType\",\r\n    \"message\": \"The resource type could not be found in the namespace 'Microsoft.App' for api version '2021-03-01'.\"\r\n  }\r\n}"}]}}
```

full template...

```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "location": {
            "type": "string"
        },
        "environmentName": {
            "type": "string"
        },
        "log_aw_id": {
            "type": "string"
        },
        "log_aw_key": {
            "type": "string"
        }
    },
    "resources": [
        {
            "name": "[parameters('environmentName')]",
            "location": "[parameters('location')]",
            "properties": {
                "internalLoadBalancerEnabled": false,
                "appLogsConfiguration": {
                    "destination": "log-analytics",
                    "logAnalyticsConfiguration": {
                        "customerId": "[parameters('log_aw_id')]",
                        "sharedKey": "[parameters('log_aw_key')]"
                    }
                },
                "type": "managed"
            },
            "apiVersion": "2021-03-01",
            "type": "Microsoft.App/kubeEnvironments"
        }
    ]
}
```

**Comment by ruslany** [comment]
@adeturner - the new CLI is not ready yet. We expect it to be ready this week. Meanwhile I've added example bicep and ARM templates for creating environments and containerapps in Microsoft.App namespace:

ARM template: https://github.com/microsoft/azure-container-apps/blob/main/docs/templates/arm/main.json
Bicep template: https://github.com/microsoft/azure-container-apps/blob/main/docs/templates/bicep/main.bicep (thanks to @KaiWalter for his test templates, which I used as a reference)

Please note that there is no `KubeEnvironment` resource in Microsoft.App namespace. Instead there is `ManagedEnvironment`.

**Comment by ruslany** [comment]
@adeturner - you can also use this link (https://aka.ms/microsoftappportal) to use Azure Portal to create environments and containerapps in the new namespace.

**Comment by joergjo** [comment]
Is there also a ready to run Bicep sample for a `Microsoft.App/managedEnvironments/daprComponents` available? I'm getting a nondescript BadRequest error with a status Failed/NotFound while trying to deploy this component in North EU and West EU:

```
resource imageStoreComponent 'Microsoft.App/managedEnvironments/daprComponents@2022-01-01-preview' = {
  name: 'imageStore'
  parent: environment
  properties: {
    componentType: 'bindings.azure.blobstorage'
    version: 'v1' 
    metadata: [
      {
        name: 'storageAccount'
        value: storageAccountName
      }
      {
        name: 'storageAccessKey'
        secretRef: 'storage-key'
      }
      {
        name: 'container'
        value: containerName
      }
      {
        name: 'decodeBase64'
        value: 'true'
      }
    ]
    secrets: [
      {
        name: 'storage-key'
        value: storageAccount.listKeys().keys[0].value
      }
    ] 
  }
}
```

**Comment by joergjo** [comment]
> Hello @clarenceb - the daprcomponents API implementation is not deployed to production yet. We expect it to be available in a week.

Ah, I guess that's what tripping me. 

@ruslany Has this been deployed by now?

**Comment by ruslany** [comment]
@joergjo - it hasn't been yet. We are trying to finish deployment this week.

**Comment by ffrankozz** [comment]
Hi,
I cannot create Azure Container Apps environment and hence no Container Apps in N.Europe and West Europe; I was able to do that   two weeks ago. I tried Central Canada just for the sake of it and I am allowed there. Is this problem due to the Microsoft.App migration?

**Comment by ruslany** [comment]
@ffrankozz - if you are getting an error "Creation or update of environments and container apps in Microsoft.Web namespace are no longer supported. Please use Microsoft.App namespace instead. More information available via aka.ms/microsoftappmigration" then use this Azure Portal link to create a new environment:

https://aka.ms/microsoftappportal

Also you can create environments using ARM or Bicep templates. Examples here:

ARM template: https://github.com/microsoft/azure-container-apps/blob/main/docs/templates/arm/main.json
Bicep template: https://github.com/microsoft/azure-container-apps/blob/main/docs/templates/bicep/main.bicep

**Comment by ffrankozz** [comment]
@ruslany- No,  the error I receive is as if my subscription is  not allowed to create container apps environment in those locations. I wonder why should that be? If I can create it in "canadacentral"  location, I would also be able to do it in N.Europe which I was able to do a few weeks ago..  I am still  using Micrososft.Web.  
Another thing is: I wonder why I do net get the error you mention.

**Comment by ruslany** [comment]
@ffrankozz - yes, try to re-register your subscription with Microsoft.App:

`az provider register -n Microsoft.App`

**Comment by ffrankozz** [comment]
@ruslany- Ok. That is  the case if I use the azure cli. What about me not being able to do it in the portal directly. What might be the reason? Have you already changed the portal's  container apps workings according to the Microsoft.App namespace. If so, may I ask, when it has become effective in N.Europe or West Europe locations? I want to figure out whether when I tried N.Europe in the portal 12 hours ago, Micrososft.App was in action.

**Comment by ruslany** [comment]
@ffrankozz - portal currently uses Microsoft.App when creating app only if the corresponding environment is already in Microsoft.App. Newly created environments are still attempted to be created in Microsoft.Web by default. To override that you can launch Azure portal by using this link: https://aka.ms/microsoftappportal. With that link the new environments will be created in Microsoft.App.

The new namespace is already effective in North Europe and West Europe. What is the exact error that you get when trying to create an environment in those regions?

**Comment by ffrankozz** [comment]
@ruslany- The error I get is: you subscription id such an such  is not allowed to build environments in that region... Might not be the exact words but something like that.

**Comment by ruslany** [comment]
@ffrankozz - there was a bug in the North Europe region that may have caused this error. It has been fixed today. Can you try again to create an environment in that region?

**Comment by ffrankozz** [comment]
@ruslany- Yes, I can create it in the N.Europe region now directly in the portal. I have not tried it via azure cli yet.  However, the environment resource  is still using Micorsoft.Web/kubeEnvironments as I notice.

**Comment by timoknapp** [comment]
> @timoknapp - please hold off on testing the dapr functionality. We are making some changes to the API specification and implementation. We'll update the announcement with more details shortly.

Hi @ruslany , are there any news on this? Dapr still seems to be broken as I am not able to connect to the Dapr host with my application running on the new API of container apps:

```
3/21/2022, 10:11:51.291 AM	io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection refused: localhost/127.0.0.1:3500	
3/21/2022, 10:11:51.291 AM	Caused by: java.net.ConnectException: finishConnect(..) failed: Connection refused
```

I also tried provisioning the container apps env with the old `Microsoft.Web` API but it does not work anymore. I guess you already discarded the old API, did you?

```
Location: West Europe
Tracking ID: 4e1ae416-6136-4565-b86b-64061714cafc
Service Request ID: acc2c067-a032-4274-b2dc-b88fc50521df

{
    "status": "Failed",
    "error": {
        "code": "InternalServerError",
        "message": "There was an unexpected InternalServerError.  Please try again later.  x-ms-correlation-request-id: e9b789fa-f51a-4458-a2cb-257ad96ff090"
    }
}
```

Thanks and regards,
Timo

**Comment by onionhammer** [comment]
Is this live on all subscriptions? I get a 'bad request' when I try to configure managed environment level dapr components

**Comment by ruslany** [comment]
@timoknapp , @onionhammer - the dapr changes are not in all regions yet. Our release got delayed. We are working on deploying the change this week.

**Comment by ruslany** [comment]
> @ruslany- Yes, I can create it in the N.Europe region now directly in the portal. I have not tried it via azure cli yet. However, the environment resource is still using Micorsoft.Web/kubeEnvironments as I notice.

@ffrankozz - did you try to create the environment using Azure portal following this link https://aka.ms/microsoftappportal?

**Comment by ffrankozz** [comment]
@ruslany- I have just tried to create one with the link you have given above. Yes, I can create an  app in the portal based on Microsoft.App namespace in N. Europe location. Are  dapr changes already  in effect in N or  W.  Europe? I  have created a microservices app for teaching purposes based on Micrososft.Web and Dapr. What happens to it now? Should I update it to use Micrososft.App.

**Comment by ruslany** [comment]
@ffrankozz - the dapr changes should be in W Europe this week. The apps and environments that you have already created in Microsoft.Web will be migrated to Microsoft.App. We plan to do a force migration at the end of March.

**Comment by Kandersen-git** [comment]
@ruslany I use the https://aka.ms/microsoftappportal to creat an Azure Container App and I have registereded the resource provider Microsoft.App, but it fails with "Managed environment failed to initialize". The resource group MC_ is created with its resources and the app environment. If I then try to add a container app, the hello world app, it fails with "code":"90015","message":"The environment 'cen-elk-01' has not been provisioned successfully." Resources are created in West Europe. Should this work now?

**Comment by ruslany** [comment]
@Kandersen-git - I see that this environment has been created in the Microsoft.App namespace, but its provisioning state is "Failed". We are investigating what caused it to fail.

**Comment by ruslany** [comment]
@Kandersen-git - it is possible that the environment failed to initialize because it cannot reach the api server due to network configuration. Please check the NSG rules used in the VNET for this environment to make sure it has allow rules as described here: https://github.com/microsoft/azure-container-apps/wiki/Lock-down-VNET-with-Network-Security-Groups-and-Firewall

**Comment by ruslany** [comment]
@joergjo , @clarenceb, @timoknapp , @onionhammer - the Dapr Components change has been deployed in all regions.

**Comment by timoknapp** [comment]
> @joergjo , @clarenceb, @timoknapp , @onionhammer - the Dapr Components change has been deployed in all regions.

@ruslany - I updated my bicep templates and now Dapr works just fine! Even the new probes are working as expected. This is amazing - thank you :)

**Comment by KaiWalter** [comment]
> > @joergjo , @clarenceb, @timoknapp , @onionhammer - the Dapr Components change has been deployed in all regions.
> 
> @ruslany - I updated my bicep templates and now Dapr works just fine! Even the new probes are working as expected. This is amazing - thank you :)

same here

**Comment by ffrankozz** [comment]
@ruslany- with reference to the bicep file example you have provided above showing the usage of the  Microsoft.App namespace:  In the sample you have defined a daprComponent and a containerApp... How are we supposed to link them? Is it going to be as before with a components section within the containerapp bicep file?  If the answer is yes, that means there is only the namespace change in  daprcomponent bicep file definition.

**Comment by ruslany** [comment]
@ffrankozz - the bicep file has been updated to show the usage. Note the appId parameter which is used in to in the containerapp.properties.configuration.dapr section and in daprcomponent.properties.scopes section.

**Comment by mmckechney** [comment]
When can we expect an updated version of the `az containerapp` extension that will support the `Microsoft.App` resource provider?

**Comment by ffrankozz** [comment]
@ruslany- So If I understand correctly, ContainerApp's properties.configuration section is where we can reference  the dapr components(before we would do that in template section) and also place the dapr appId's of the microservices(those using the component)  in the scopes array of the dapr component resource definition.

**Comment by kendallroden** [comment]
Hi @ffrankozz - there is no concept of referencing dapr components in the container app with this change. The properties.configuration section is where you will set dapr properties for the container app i.e. enabled, App Port, App Protocol, App Id

The components are deployed at the environment level. Any dapr-enabled container app can access these components. You are able to scope them by adding a container app's App Id value to the component. 

A quick example: 
I have an environment where I deployed an azure blob state store component called "statestore" with no scopes

I have two dapr-enabled containers apps: 
ContainerAppA with dapr app id of A
ContainerAppB with dapr app id of B 

By default, the dapr sidecar for both apps would load the state store component called "statestore" at runtime 
In order to restrict the runtime behavior, I can add a scope to "statestore" and include dapr app id A 
Once this is done, ContainerAppB with dapr app id of B would not load that component at runtime 

There is no reference to the component in the container app spec (essentially, our implementation is now closely aligned with the behavior of dapr in a kubernetes namespace)

**Comment by kendallroden** [comment]
> When can we expect an updated version of the `az containerapp` extension that will support the `Microsoft.App` resource provider?

sometime in the next day or two

**Comment by ffrankozz** [comment]
Hi @kendallroden Thank you very much for the explanation. It is all clear now. In addition to what you said, it also seems to match the local development style  where we have dapr components folder full of components which may have dapr app scopes and,  at runtime, any microservice which the component has scope of,  loads the component.

**Comment by onionhammer** [comment]
It would be cool to be able to load dapr configuration files within a bicep file so you can keep the dapr configuration in yaml instead of dapr

**Comment by anthonychu** [comment]
Hey everyone. Made some updates to this announcement. The Microsoft.App compatible CLI is available (see #152). The Azure portal is also now updated to create Microsoft.App resources by default.

**Comment by Kandersen-git** [comment]
> @Kandersen-git - it is possible that the environment failed to initialize because it cannot reach the api server due to network configuration. Please check the NSG rules used in the VNET for this environment to make sure it has allow rules as described here: https://github.com/microsoft/azure-container-apps/wiki/Lock-down-VNET-with-Network-Security-Groups-and-Firewall

**Comment by forki** [comment]
so looks since of today the old stuff doesn't work anymore. How great. Or not

**Comment by ruslany** [comment]
@forki - the migration of the resources from Microsoft.Web to Microsoft.App namespace is currently in progress. The environments and apps from the Microsoft.Web namespace should become available in the Microsoft.App namespace. Let me know if that is not the case for apps in your subscription.

**Comment by forki** [comment]
@ruslany the problem is not that the new one doesn't exist. The problem is that the old one does not work anymore. And the change comes at the time whene MS changed root certificates for our *.azurewebsites..net sites which brought down A LOT of IoT devices.

**Comment by chrisreddington** [comment]
> > @joergjo , @clarenceb, @timoknapp , @onionhammer - the Dapr Components change has been deployed in all regions.
> 
> @ruslany - I updated my bicep templates and now Dapr works just fine! Even the new probes are working as expected. This is amazing - thank you :)

Likewise, I've just deployed using the new Resource Providers, and refactored some of my old Bicep Files. All looks to work great :)

**Comment by askaribragimov** [comment]
Found this bug by trying to attach a container app workspace to a log  analytics workspace.

```
az extension add --name containerapp --upgrade
>  Extension 'containerapp' 0.3.20 is already installed.
az provider register -n Microsoft.App
az monitor log-analytics workspace create --resource-group "$(resourceGroup)"  --workspace-name "$(envName)-logworkspace" 
az containerapp env create  --name "$(containerAppsEnvironment)"  --resource-group "$(resourceGroup)"  --location "West Europe" --logs-workspace-id "$WORKSPACE_ID" --logs-workspace-key $WORKSPACE_SHARED_KEY

ERROR: (InvalidRequestParameterWithDetails) LogAnalyticsConfiguration.CustomerId is invalid.  CustomerId must be a GUID without additional whiteSpace.
```

Documentation of `az containerapp env create` has no notion of Customer ID: https://learn.microsoft.com/en-us/cli/azure/containerapp/env?view=azure-cli-latest#az-containerapp-env-create

It is very hard to comprehend, why documentation has no clue whatsoever as to why this error happens and a lot of developer resources are channeled to spend time figuring yet another puzzler increasing time to market and lowering ROI for the company that is trying to use the technology.

**Comment by willvelida** [comment]
Hey @askaribragimov, I've just tried to provision an environment with Log Analytics and I got it working. Here's what I did for reference:

```
$RESOURCE_GROUP_NAME="rg-acaazcli"
$LOCATION="australiaeast"

az group create --name $RESOURCE_GROUP_NAME --location $LOCATION

$WORKSPACE_NAME="lawwvacaazcli"
az monitor log-analytics workspace create --resource-group $RESOURCE_GROUP --workspace-name $WORKSPACE_NAME

$WORKSPACE_ID=(az monitor log-analytics workspace show --resource-group $RESOURCE_GROUP --workspace-name $WORKSPACE_NAME --query customerId -o tsv)

$WORKSPACE_SHARED_KEY=(az monitor log-analytics workspace get-shared-keys --resource-group $RESOURCE_GROUP --workspace-name $WORKSPACE_NAME --query "primarySharedKey" -o tsv)

$CONTAINER_APP_ENVIRONMENT_NAME="envwvacaazcli"
az containerapp env create --name $CONTAINER_APP_ENVIRONMENT_NAME --resource-group $RESOURCE_GROUP --location australiaeast --logs-workspace-id $WORKSPACE_ID --logs-workspace-key $WORKSPACE_SHARED_KEY
```

Can I ask you how you are extracting the value your $WORKSPACE_ID? That might be causing the issue.

**Comment by askaribragimov** [comment]
@willvelida `--query customerId` was the key, I used `--query id`.

Why the hell workspace ID is also customer ID? Microsoft never ceases to name same thing differently. They should really take AWS example where such thing does not seem to be of issue.

**Comment by willvelida** [comment]
@askaribragimov resource versions do change over time, so issues like this can occur. Sorry about the inconvenience caused and glad to hear that querying customerId worked for you.

@anthonychu - hopefully you're the right person to ask, but the parameter ```--logs-workspace-id``` is a bit confusing when it's the ```customerId``` that should be supplied. Is there anyway we can improve the error message to state that the customer Id should be provided, or can we improve the doc on ```az containerapp env create``` to make this clearer?

**Comment by askaribragimov** [comment]
@willvelida thank you for the prompt reaction! Apologies that I sound frustrated, it is really confusing what's going on there. 
I would suggest renaming the parameter to `--logs-workspace-customer-id` that would harmonize the naming with log workspace terminology. Note that currently you ask for `workspace id` but if it is wrong, an error message talks about Customer ID, so within `az containerapp create` the terminology is not currently consistent between input parameters/docs and the errors it prints.

**Comment by vienleidl** [comment]
> @willvelida thank you for the prompt reaction! Apologies that I sound frustrated, it is really confusing what's going on there. I would suggest renaming the parameter to `--logs-workspace-customer-id` that would harmonize the naming with log workspace terminology. Note that currently you ask for `workspace id` but if it is wrong, an error message talks about Customer ID, so within `az containerapp create` the terminology is not currently consistent between input parameters/docs and the errors it prints.

I got the same issue. Thanks @willvelida for your reference command lines, adding 2 `-o tsv` for getting the correct Workspace Id and Workspace Key. It's caused by the output strings have double-quotes.
